### PR TITLE
refactor: faster from_pdgid

### LINKED ISSUE #348: Vectorised lookup
Hi there,

Love the project. Just wondering whether there is a way to use something like `from_pdgid` with an iterable of integers. Instantiating a new particle object for each is very expensive, and if I'm just trying to quickly look up a restricted selection of properties en masse it isn't very practical for my use case.

Kind regards,
Jacan.

### PR DESCRIPTION
**Pull request by henryiii**
This should improve the performance of `from_pdgid` by using a hashed lookup instead of a linear search. Closes #348.

### CODE REVIEW

**Comment by eduardo-rodrigues** [comment]
Hi @henryiii, this seems like to totally viable and alternative option to me. But have you run some timing tests for standard workflows and also loading explicitly a table (say the particles one) and then a second one (say the nuclei) to ensure we're good? Just being maniac as he had twice issues with explosion of time to load.

**Comment by henryiii** [comment]
Before:
```
------------------------------------------------------------------------------------- benchmark: 2 tests -------------------------------------------------------------------------------------
Name (time in ms)                Min                 Max               Mean            StdDev             Median               IQR            Outliers       OPS            Rounds  Iterations
----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
test_load_particle_table      7.3516 (1.0)       21.5053 (1.0)       7.9696 (1.0)      1.3971 (1.0)       7.6958 (1.0)      0.1936 (1.0)          5;12  125.4763 (1.0)         108           1
test_load_nuclei             87.5544 (11.91)    116.3791 (5.41)     95.5472 (11.99)    7.9544 (5.69)     93.5607 (12.16)    3.3593 (17.35)         2;2   10.4660 (0.08)         11           1
----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
```

After:
```
------------------------------------------------------------------------------------- benchmark: 2 tests ------------------------------------------------------------------------------------
Name (time in ms)                Min                Max               Mean            StdDev             Median               IQR            Outliers       OPS            Rounds  Iterations
---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
test_load_particle_table      5.9915 (1.0)      19.5842 (1.0)       6.3464 (1.0)      1.1441 (1.0)       6.1928 (1.0)      0.1676 (1.0)          4;11  157.5703 (1.0)         146           1
test_load_nuclei             58.0820 (9.69)     81.8952 (4.18)     61.9745 (9.77)     7.6346 (6.67)     59.1297 (9.55)     1.5327 (9.14)          2;2   16.1357 (0.10)         16           1
---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
```

File: tests/particle/test_performance.py [hunk 1]
```diff
@@ -0,0 +1,17 @@
+# -*- coding: utf-8 -*-
+
+from particle import Particle, data
+
+
+def test_load_particle_table(benchmark):
+    benchmark(Particle.load_table, data.basepath / "particle2018.csv")
```
Comment (eduardo-rodrigues) [inline tests/particle/test_performance.py:7]:
```suggestion
    benchmark(Particle.load_table, data.basepath / "particle2021.csv")
```

Comment (eduardo-rodrigues) [reply tests/particle/test_performance.py:7]: Just so that we use the latest and more relevant file. 2018 is "ancient".

**Comment by eduardo-rodrigues** [comment]
Nice to see these benchmarks 👍!

**Comment by henryiii** [comment]
I'll be fixing the third one (the lookup one) before merging. It is checking a particle near the front of the list so the performance is similar. A larger number would be more interesting. I'll probably have it check a loop over all numbers.

**Comment by eduardo-rodrigues** [comment]
Sounds good. Approving module the couple of updates in the pipeline (see the trivial suggestion above).

**Review by eduardo-rodrigues — APPROVED** [review]

File: setup.cfg [hunk 2]
```diff
@@ -81,7 +84,9 @@ universal = 1
 testpaths =
     tests
 junit_family = xunit2
-addopts = '-Wd'
+addopts = --benchmark-disable -ra --showlocals --strict-config
```
Comment (eduardo-rodrigues) [inline setup.cfg:87]: So `--strict-config` is the only providing the failure in Python 2.7. I wonder if `addopts` can be dependent on the Python version as dependencies specified in this file?

Comment (henryiii) [reply setup.cfg:87]: We will just have to leave it off until we drop 2.7.

Comment (henryiii) [reply setup.cfg:87]: (It's stuck with an old PyTest)

**Comment by henryiii** [comment]
Slightly nicer, and doing a bit more for the append. New from PDGID section, roughly doubles lookup performance, and increases the rest a bit.

Old:
```
------------------------------------------------------------------------------------- benchmark: 3 tests -------------------------------------------------------------------------------------
Name (time in ms)                Min                 Max               Mean            StdDev             Median               IQR            Outliers       OPS            Rounds  Iterations
----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
test_load_particle_table      7.4517 (1.0)       20.8920 (1.83)      8.1380 (1.0)      1.2724 (4.36)      7.9353 (1.0)      0.1741 (1.0)          4;19  122.8804 (1.0)         118           1
test_from_pdgid               9.6907 (1.30)      11.4370 (1.0)      10.4456 (1.28)     0.2917 (1.0)      10.4813 (1.32)     0.3427 (1.97)         27;3   95.7337 (0.78)         95           1
test_load_nuclei_append      82.6308 (11.09)    102.7947 (8.99)     87.4490 (10.75)    5.4610 (18.72)    84.9447 (10.70)    4.0507 (23.26)         1;1   11.4352 (0.09)         12           1
----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
```

New:
```
------------------------------------------------------------------------------------- benchmark: 3 tests ------------------------------------------------------------------------------------
Name (time in ms)                Min                Max               Mean            StdDev             Median               IQR            Outliers       OPS            Rounds  Iterations
---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
test_from_pdgid               4.5469 (1.0)       7.0901 (1.0)       4.8089 (1.0)      0.2603 (1.0)       4.7585 (1.0)      0.1051 (1.0)         16;19  207.9494 (1.0)         207           1
test_load_particle_table      6.2856 (1.38)     18.8554 (2.66)      6.9040 (1.44)     1.1362 (4.37)      6.7140 (1.41)     0.2250 (2.14)         5;11  144.8426 (0.70)        125           1
test_load_nuclei_append      67.6541 (14.88)    86.8385 (12.25)    71.8240 (14.94)    6.2749 (24.11)    69.1568 (14.53)    2.4218 (23.03)         2;2   13.9229 (0.07)         15           1
---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
```

**Pull request merged**